File: pipeline/holdings-browser/filings.py
"""Filing dates, previous/next links, and dated local/B2 paths."""
from __future__ import annotations

import calendar
import re
from typing import Any

NO_DATA_FOUND = "No Data Found"
LINK_KEYS = ("as_of", "href", "message")

_MONTHS = {
    "jan": 1,
    "january": 1,
    "feb": 2,
    "february": 2,
    "mar": 3,
    "march": 3,
    "apr": 4,
    "april": 4,
    "may": 5,
    "jun": 6,
    "june": 6,
    "jul": 7,
    "july": 7,
    "aug": 8,
    "august": 8,
    "sep": 9,
    "sept": 9,
    "september": 9,
    "oct": 10,
    "october": 10,
    "nov": 11,
    "november": 11,
    "dec": 12,
    "december": 12,
}
_FORTNIGHTLY_RE = re.compile(
    r"\b(debt|liquid|overnight|money\s*market|gilt|credit\s*risk|floater|"
    r"ultra\s*short|ultrashort|low\s*duration|short\s*duration|medium\s*duration|"
    r"corporate\s*bond|banking\s*(?:and|&)\s*psu|dynamic\s*bond)\b",
    re.I,
)


def format_date(year: int, month: int, day: int) -> str:
    return f"{year:04d}-{month:02d}-{day:02d}"


def last_day_of_month(year: int, month: int) -> int:
    return calendar.monthrange(year, month)[1]


def normalize_as_of(raw: Any) -> str:
    s = str(raw or "").strip()
    if not s:
        return ""
    m = re.match(r"^(\d{4})-(\d{1,2})-(\d{1,2})$", s)
    if m:
        y, mo, d = int(m.group(1)), int(m.group(2)), int(m.group(3))
        if mo < 1 or mo > 12:
            return ""
        dim = last_day_of_month(y, mo)
        if d < 1 or d > dim:
            return ""
        return format_date(y, mo, d)
    m = re.match(r"^(\d{4})-(\d{1,2})$", s)
    if m:
        y, mo = int(m.group(1)), int(m.group(2))
        if mo < 1 or mo > 12:
            return ""
        return format_date(y, mo, last_day_of_month(y, mo))
    m = re.match(r"^(\d{1,2})[./-](\d{1,2})[./-](\d{4})$", s)
    if m:
        d, mo, y = int(m.group(1)), int(m.group(2)), int(m.group(3))
        if mo < 1 or mo > 12:
            return ""
        dim = last_day_of_month(y, mo)
        if d < 1 or d > dim:
            return ""
        return format_date(y, mo, d)
    m = re.match(r"^(\d{1,2})[./\s-]+([A-Za-z]+)[./\s,-]+(\d{4})$", s)
    if m:
        mo = _MONTHS.get(m.group(2).lower())
        if not mo:
            return ""
        d, y = int(m.group(1)), int(m.group(3))
        dim = last_day_of_month(y, mo)
        if d < 1 or d > dim:
            return ""
        return format_date(y, mo, d)
    return ""


def is_fortnightly(scheme: dict[str, Any] | None, meta: dict[str, Any] | None = None) -> bool:
    scheme = scheme or {}
    meta = meta or {}
    dtype = str(meta.get("disclosure_type") or scheme.get("disclosure_type") or "").lower()
    if dtype == "fortnightly":
        return True
    if dtype == "monthly":
        return False
    blob = f"{scheme.get('category') or ''} {scheme.get('name') or ''}"
    return bool(_FORTNIGHTLY_RE.search(blob))


def previous_filing_date(as_of: str, fortnightly: bool) -> str:
    y, m, d = (int(x) for x in as_of.split("-"))
    if fortnightly:
        if d > 15:
            return format_date(y, m, 15)
        if m == 1:
            return format_date(y - 1, 12, last_day_of_month(y - 1, 12))
        return format_date(y, m - 1, last_day_of_month(y, m - 1))
    if m == 1:
        return format_date(y - 1, 12, last_day_of_month(y - 1, 12))
    return format_date(y, m - 1, last_day_of_month(y, m - 1))


def next_filing_date(as_of: str, fortnightly: bool) -> str:
    y, m, d = (int(x) for x in as_of.split("-"))
    eom = last_day_of_month(y, m)
    if fortnightly:
        if d < 15:
            return format_date(y, m, 15)
        if d < eom:
            return format_date(y, m, eom)
        if m == 12:
            return format_date(y + 1, 1, 15)
        return format_date(y, m + 1, 15)
    if d < eom:
        return format_date(y, m, eom)
    if m == 12:
        return format_date(y + 1, 1, last_day_of_month(y + 1, 1))
    return format_date(y, m + 1, last_day_of_month(y, m + 1))


def dated_b2_key(latest_key: str, as_of: str) -> str:
    if not latest_key or not as_of:
        return ""
    if "/holdings/latest/" in latest_key:
        return latest_key.replace("/holdings/latest/", f"/holdings/{as_of}/")
    return re.sub(r"/holdings/\d{4}-\d{2}-\d{2}/", f"/holdings/{as_of}/", latest_key)


def amfi_href(origin: str, code: str, as_of: str) -> str:
    return f"{origin.rstrip('/')}/api/amfi/{code}?as_of={as_of}"


def _link(as_of: str, href: str, available: bool) -> dict[str, Any]:
    return {"as_of": as_of, "href": href, "message": None if available else NO_DATA_FOUND}


def filing_links(
    *,
    origin: str,
    code: str,
    as_of: str,
    previous_as_of: str,
    next_as_of: str,
    previous_available: bool,
    next_available: bool,
) -> dict[str, Any]:
    return {
        "self": _link(as_of, amfi_href(origin, code, as_of), True),
        "previous": _link(previous_as_of, amfi_href(origin, code, previous_as_of), previous_available),
        "next": _link(next_as_of, amfi_href(origin, code, next_as_of), next_available),
    }


def no_data_payload(
    scheme: dict[str, Any] | None,
    as_of: str | None,
    links: dict[str, Any] | None = None,
) -> dict[str, Any]:
    scheme = scheme or {}
    return {
        "error": NO_DATA_FOUND,
        "amfi_code": scheme.get("amfi_code"),
        "as_of": as_of,
        "scheme": {
            "amfi_code": scheme.get("amfi_code"),
            "name": scheme.get("name"),
            "amc_name": scheme.get("amc_name"),
            "parent_name": scheme.get("parent_name"),
        },
        "links": links,
    }


def dated_local_candidates(rel: str, as_of: str) -> list[str]:
    """Swap the parsed period folder for another as-of date."""
    parts = rel.replace("\\", "/").split("/")
    if len(parts) < 6 or parts[0] != "data" or parts[1] != "parsed":
        return []
    y, m, d = (int(x) for x in as_of.split("-"))
    eom = last_day_of_month(y, m)
    periods: list[tuple[str, str]] = []
    if d == 15:
        periods.append(("fortnightly", f"{y:04d}-{m:02d}-15"))
    if d == eom:
        periods.append(("monthly", f"{y:04d}-{m:02d}"))
        periods.append(("fortnightly", f"{y:04d}-{m:02d}"))
    periods.append((parts[2], as_of))
    out: list[str] = []
    seen: set[str] = set()
    rest = "/".join(parts[4:])
    for dtype, period in periods:
        cand = f"data/parsed/{dtype}/{period}/{rest}"
        if cand not in seen:
            seen.add(cand)
            out.append(cand)
    return out

File: pipeline/holdings-browser/api/_filings.js
/** Filing dates, B2 key layout, and previous/next links. */

export const NO_DATA_FOUND = "No Data Found";

export const LINK_KEYS = ["as_of", "href", "message"];

const MONTHS = {
  jan: 1,
  january: 1,
  feb: 2,
  february: 2,
  mar: 3,
  march: 3,
  apr: 4,
  april: 4,
  may: 5,
  jun: 6,
  june: 6,
  jul: 7,
  july: 7,
  aug: 8,
  august: 8,
  sep: 9,
  sept: 9,
  september: 9,
  oct: 10,
  october: 10,
  nov: 11,
  november: 11,
  dec: 12,
  december: 12,
};

const FORTNIGHTLY_RE =
  /\b(debt|liquid|overnight|money\s*market|gilt|credit\s*risk|floater|ultra\s*short|ultrashort|low\s*duration|short\s*duration|medium\s*duration|corporate\s*bond|banking\s*(?:and|&)\s*psu|dynamic\s*bond)\b/i;

export function lastDayOfMonth(year, month) {
  return new Date(Date.UTC(year, month, 0)).getUTCDate();
}

export function formatDate(year, month, day) {
  return `${year}-${String(month).padStart(2, "0")}-${String(day).padStart(2, "0")}`;
}

function parts(asOf) {
  const [y, m, d] = String(asOf).split("-").map(Number);
  return { y, m, d };
}

export function normalizeAsOf(raw) {
  const s = String(raw || "").trim();
  if (!s) return "";
  let m = s.match(/^(\d{4})-(\d{1,2})-(\d{1,2})$/);
  if (m) {
    const y = Number(m[1]);
    const mo = Number(m[2]);
    const d = Number(m[3]);
    if (mo < 1 || mo > 12) return "";
    const dim = lastDayOfMonth(y, mo);
    if (d < 1 || d > dim) return "";
    return formatDate(y, mo, d);
  }
  m = s.match(/^(\d{4})-(\d{1,2})$/);
  if (m) {
    const y = Number(m[1]);
    const mo = Number(m[2]);
    if (mo < 1 || mo > 12) return "";
    return formatDate(y, mo, lastDayOfMonth(y, mo));
  }
  m = s.match(/^(\d{1,2})[\/.\-](\d{1,2})[\/.\-](\d{4})$/);
  if (m) {
    const d = Number(m[1]);
    const mo = Number(m[2]);
    const y = Number(m[3]);
    if (mo < 1 || mo > 12) return "";
    const dim = lastDayOfMonth(y, mo);
    if (d < 1 || d > dim) return "";
    return formatDate(y, mo, d);
  }
  m = s.match(/^(\d{1,2})[\/.\-\s]+([A-Za-z]+)[\/.\-\s,]+(\d{4})$/);
  if (m) {
    const mo = MONTHS[m[2].toLowerCase()];
    if (!mo) return "";
    const d = Number(m[1]);
    const y = Number(m[3]);
    const dim = lastDayOfMonth(y, mo);
    if (d < 1 || d > dim) return "";
    return formatDate(y, mo, d);
  }
  return "";
}

export function isFortnightly(scheme, meta) {
  const dtype = String(meta?.disclosure_type || scheme?.disclosure_type || "").toLowerCase();
  if (dtype === "fortnightly") return true;
  if (dtype === "monthly") return false;
  const blob = `${scheme?.category || ""} ${scheme?.name || ""}`;
  return FORTNIGHTLY_RE.test(blob);
}

export function previousFilingDate(asOf, fortnightly) {
  const { y, m, d } = parts(asOf);
  if (fortnightly) {
    if (d > 15) return formatDate(y, m, 15);
    const py = m === 1 ? y - 1 : y;
    const pm = m === 1 ? 12 : m - 1;
    return formatDate(py, pm, lastDayOfMonth(py, pm));
  }
  const py = m === 1 ? y - 1 : y;
  const pm = m === 1 ? 12 : m - 1;
  return formatDate(py, pm, lastDayOfMonth(py, pm));
}

export function nextFilingDate(asOf, fortnightly) {
  const { y, m, d } = parts(asOf);
  const eom = lastDayOfMonth(y, m);
  if (fortnightly) {
    if (d < 15) return formatDate(y, m, 15);
    if (d < eom) return formatDate(y, m, eom);
    const ny = m === 12 ? y + 1 : y;
    const nm = m === 12 ? 1 : m + 1;
    return formatDate(ny, nm, 15);
  }
  if (d < eom) return formatDate(y, m, eom);
  const ny = m === 12 ? y + 1 : y;
  const nm = m === 12 ? 1 : m + 1;
  return formatDate(ny, nm, lastDayOfMonth(ny, nm));
}

export function datedB2Key(latestKey, asOf) {
  if (!latestKey || !asOf) return "";
  if (latestKey.includes("/holdings/latest/")) {
    return latestKey.replace("/holdings/latest/", `/holdings/${asOf}/`);
  }
  return latestKey.replace(
    /\/holdings\/\d{4}-\d{2}-\d{2}\//,
    `/holdings/${asOf}/`,
  );
}

export function publicOrigin(req) {
  const host =
    req?.headers?.["x-forwarded-host"] ||
    req?.headers?.host ||
    "fund-holdings-browser.vercel.app";
  const proto = req?.headers?.["x-forwarded-proto"] || "https";
  return `${proto}://${host}`;
}

export function amfiHref(req, code, asOf) {
  return `${publicOrigin(req)}/api/amfi/${encodeURIComponent(code)}?as_of=${encodeURIComponent(asOf)}`;
}

function link(asOf, href, available) {
  return {
    as_of: asOf,
    href,
    message: available ? null : NO_DATA_FOUND,
  };
}

export function filingLinks({ req, code, asOf, previousAsOf, nextAsOf, previousAvailable, nextAvailable }) {
  return {
    self: link(asOf, amfiHref(req, code, asOf), true),
    previous: link(previousAsOf, amfiHref(req, code, previousAsOf), previousAvailable),
    next: link(nextAsOf, amfiHref(req, code, nextAsOf), nextAvailable),
  };
}

export function noDataPayload({ scheme, asOf, links }) {
  return {
    error: NO_DATA_FOUND,
    amfi_code: scheme?.amfi_code || null,
    as_of: asOf || null,
    scheme: scheme
      ? {
          amfi_code: scheme.amfi_code || null,
          name: scheme.name || null,
          amc_name: scheme.amc_name || null,
          parent_name: scheme.parent_name || null,
        }
      : null,
    links: links || null,
  };
}
